### ISSUE TITLE
Trying to selfupdate from Composer to Composer2

### ISSUE THREAD

**Issue by richb201** [composer/composer#9866]
richb201@richb201-XPS-13-9370:/opt/docker-substantiator2/app/vendor/sendpulse-rest-api-php-master/src$ composer selfupdate
Updating to version 2.0.13 (stable channel).
   Downloading (100%)         

I get this error:
                                                                                                        
  [ErrorException]                                                                                      
  rename(/home/richb201/.cache/composer/composer-temp.phar,/usr/local/bin/composer): Permission denied

**Comment by kubawerlos** [comment]
Use `sudo` :)

Or fix permissions of `/usr/local/bin/composer`.

**Comment by richb201** [comment]
perfect. Thank you!!!

**Comment by richb201** [comment]
When I try running I now get:
***@***.***XPS-13-9370:/opt/docker-substantiator2/app$ composer
require sendpulse/rest-api
Using version ^1.0 for sendpulse/rest-api
./composer.json has been updated
Running composer update sendpulse/rest-api
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - aws/aws-sdk-php is locked to version 3.145.1 and an update of this
package was not requested.
    - aws/aws-sdk-php 3.145.1 requires ext-simplexml * -> it is missing
from your system. Install or enable PHP's simplexml extension.
  Problem 2
    - php-http/httplug-bundle is locked to version 1.18.0 and an update of
this package was not requested.
    - php-http/httplug-bundle 1.18.0 requires php ^7.2 -> your php version
(5.6.40) does not satisfy that requirement.
  Problem 3
    - fusionauth/fusionauth-client is locked to version 1.19.0 and an
update of this package was not requested.
    - fusionauth/fusionauth-client 1.19.0 requires ext-curl * -> it is
missing from your system. Install or enable PHP's curl extension.

To enable extensions, verify that they are enabled in your .ini files:
    - /etc/php/5.6/cli/php.ini
    - /etc/php/5.6/cli/conf.d/10-mysqlnd.ini
    - /etc/php/5.6/cli/conf.d/10-opcache.ini
    - /etc/php/5.6/cli/conf.d/10-pdo.ini
    - /etc/php/5.6/cli/conf.d/20-calendar.ini
    - /etc/php/5.6/cli/conf.d/20-ctype.ini
    - /etc/php/5.6/cli/conf.d/20-exif.ini
    - /etc/php/5.6/cli/conf.d/20-fileinfo.ini
    - /etc/php/5.6/cli/conf.d/20-ftp.ini
    - /etc/php/5.6/cli/conf.d/20-gettext.ini
    - /etc/php/5.6/cli/conf.d/20-iconv.ini
    - /etc/php/5.6/cli/conf.d/20-json.ini
    - /etc/php/5.6/cli/conf.d/20-mbstring.ini
    - /etc/php/5.6/cli/conf.d/20-mcrypt.ini
    - /etc/php/5.6/cli/conf.d/20-mysql.ini
    - /etc/php/5.6/cli/conf.d/20-mysqli.ini
    - /etc/php/5.6/cli/conf.d/20-pdo_mysql.ini
    - /etc/php/5.6/cli/conf.d/20-phar.ini
    - /etc/php/5.6/cli/conf.d/20-posix.ini
    - /etc/php/5.6/cli/conf.d/20-readline.ini
    - /etc/php/5.6/cli/conf.d/20-shmop.ini
    - /etc/php/5.6/cli/conf.d/20-sockets.ini
    - /etc/php/5.6/cli/conf.d/20-sysvmsg.ini
    - /etc/php/5.6/cli/conf.d/20-sysvsem.ini
    - /etc/php/5.6/cli/conf.d/20-sysvshm.ini
    - /etc/php/5.6/cli/conf.d/20-tokenizer.ini
    - /etc/php/5.6/cli/conf.d/20-xdebug.ini
You can also run `php --ini` inside terminal to see which files are used by
PHP in CLI mode.

Installation failed, reverting ./composer.json and ./composer.lock to their
original content.

What should I do???

**Comment by kubawerlos** [comment]
Exactly what the output says:

> Install or enable PHP's curl extension.

Most likely by running `sudo apt install php5.6-curl`.

**Comment by richb201** [comment]
Thanks.
***@***.***XPS-13-9370:/opt/docker-substantiator2/app$ sudo apt
install php5.6-curl
[sudo] password for richb201:
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package php5.6-curl
E: Couldn't find any package by glob 'php5.6-curl'
E: Couldn't find any package by regex 'php5.6-curl'

**Comment by kubawerlos** [comment]
Maybe package is named `php-curl`? PHP 5.6 was sooo long ago.

**Comment by richb201** [comment]
OK, I think Composerw installed. This is what I get when I try to install
sendPulse:
***@***.***XPS-13-9370:/opt/docker-substantiator2/app/vendor/sendpulse-rest-api-php-master$
composer require sendpulse/rest-api
Using version ^1.0 for sendpulse/rest-api
Root package 'sendpulse/rest-api' cannot require itself in its composer.json

Then I took a look at the composer.json
***@***.***XPS-13-9370:/opt/docker-substantiator2/app/vendor/sendpulse-rest-api-php-master$
cat composer.json
{
  "name": "sendpulse/rest-api",
  "description": "Sendpulse REST API client library",
  "keywords": [
    "sendpulse",
    "rest",
    "api"
  ],
  "license": "MIT",
  "authors": [
    {
      "name": "Maksym Ustymenko",
      "email": ***@***.***"
    },
    {
      "name": "Alexey Moroz",
      "email": ***@***.***"
    }
  ],
  "require": {
    "php": ">=5.4.0"
  },
  "autoload": {
    "psr-4": {
      "Sendpulse\\RestApi\\": "src/"
    }
  }
}

Is this composer.json correct?

**Comment by kubawerlos** [comment]
Looks fine, why do you try you try to install the same package you are in?

**Comment by richb201** [comment]
Good question. Because when I try to run the first statement ApiClient I
get this error:
Class 'Sendpulse\RestApi\ApiClient' not found

SendPulse claimed that it was an issue with my code. But if the Composer
log I sent is AOK, then it is likely their code, IMO.

**Comment by kubawerlos** [comment]
What exactly are you trying to do? Modify `sendpulse/rest-api` or use it?

**Comment by richb201** [comment]
Just use it.

**Comment by kubawerlos** [comment]
Then why do you have `sendpulse/rest-api` in your `composer.json`?

**Comment by richb201** [comment]
I don't know. I didn't put it there. Their instructions from their github
page:

<?phprequire 'vendor/autoload.php';
// Without Composer (and instead of "require
'vendor/autoload.php'")://
require("your-path/sendpulse-rest-api-php/src/ApiInterface.php");//
require("your-path/sendpulse-rest-api-php/src/ApiClient.php");//
require("your-path/sendpulse-rest-api-php/src/Storage/TokenStorageInterface.php");//
require("your-path/sendpulse-rest-api-php/src/Storage/FileStorage.php");//
require("your-path/sendpulse-rest-api-php/src/Storage/SessionStorage.php");//
require("your-path/sendpulse-rest-api-php/src/Storage/MemcachedStorage.php");//
require("your-path/sendpulse-rest-api-php/src/Storage/MemcacheStorage.php");
use Sendpulse\RestApi\ApiClient;use Sendpulse\RestApi\Storage\FileStorage;
// API credentials from
https://login.sendpulse.com/settings/#apidefine('API_USER_ID',
'');define('API_SECRET', '');define('PATH_TO_ATTACH_FILE', __FILE__);
$SPApiClient = new ApiClient(API_USER_ID, API_SECRET, new FileStorage());

Am I best just deleting the sendpulse-rest-api-php dir and having Composer
do the clean install?

**Comment by kubawerlos** [comment]
> Am I best just deleting the sendpulse-rest-api-php dir and having Composer
do the clean install?

Yes, I would do that.

**Comment by richb201** [comment]
OK. I deleted it. What dir should I do composer require sendpulse/rest-api
in? /vendor?

**Comment by kubawerlos** [comment]
In `vendor`? Have you ever do that? Run it as always, in your project's root.

**Comment by richb201** [comment]
Here is the path to my project
/opt/docker-substantiator2/app/Vendor/sendPulse/src
I pruned sendPulse off. should I run here

$cd  /opt/docker-substantiator2/app/Vendor
$ composer require sendpulse/rest-api

**Comment by richb201** [comment]
Well I got an email from the vendor. They want me to run sudo php composer require sendpulse/rest-api
First, I had no idea that I needed to run "php" composer. Do I? 

Then in addition, when i try to run the above command I get a warning:
richb201@richb201-XPS-13-9370:/opt/docker-substantiator2/vendor/sendpulse/rest-api$ sudo composer require sendpulse/rest-api
[sudo] password for richb201: 
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Continue as root/super user [yes]? Y
Using version ^1.0 for sendpulse/rest-api
Root package 'sendpulse/rest-api' cannot require itself in its composer.json

Here is the composer.json from that dir.
{
  "name": "sendpulse/rest-api",
  "description": "Sendpulse REST API client library",
  "keywords": [
    "sendpulse",
    "rest",
    "api"
  ],
  "license": "MIT",
  "authors": [
    {
      "name": "Maksym Ustymenko",
      "email": "m.ustymenko@sendpulse.com"
    },
    {
      "name": "Alexey Moroz",
      "email": "a.moroz@sendpulse.com"
    }
  ],
  "require": {
    "php": ">=5.4.0"
  },
  "autoload": {
    "psr-4": {
      "Sendpulse\\RestApi\\": "src/"
    }
  }
}

Is there a problem with their composer.json?
